internal/stolonctl/pkg/cluster/upgrade.go

```diff
@@ -84,59 +98,297 @@ func Upgrade(ctx context.Context, config Config) error {
 		return trace.Wrap(err)
 	}
 
-	// TODO delete debugging info
-	log.Info(res)
 	if !crdclient.IsPhaseCompleted(res, crd.StolonUpgradePhaseInit) {
 		res, err = crdclient.MarkPhase(res, crd.StolonUpgradePhaseInit, crd.StolonUpgradeStatusCompleted)
 		if err != nil {
 			return trace.Wrap(err)
 		}
 	}
+	// Check status of cluster
+	res, err = upgradeControl.checkStatus(res, crd.StolonUpgradePhaseChecks)
+	if err != nil {
+		return trace.Wrap(err)
+	}
 
 	// Backup Postgres
-	if !crdclient.IsPhaseCompleted(res, crd.StolonUpgradePhaseBackupPostgres) {
-		res, err = crdclient.MarkPhase(res, crd.StolonUpgradePhaseBackupPostgres, crd.StolonUpgradeStatusInProgress)
-		if err != nil {
-			return trace.Wrap(err)
+	res, err = upgradeControl.executePhase(res, crd.StolonUpgradePhaseBackupPostgres, func() error {
```

Comment:
The `Upgrade` function feels cluttered - I wonder if you could factor some things out - like move the phases to individual functions and execute them in a loop:
```go
for _, phase := range phases {
	result, err := upgradeControl.executePhase(result, phase.name, phase.handler(upgradeControl))
	// Handle error
}
```